test/new-e2e/tests/windows/common/powershell/command_builder.go

```diff
@@ -0,0 +1,157 @@
+func (ps *powerShellCommandBuilder) InstallADDSForest(activeDirectoryDomain, passwd string) *powerShellCommandBuilder {
+	ps.cmds = append(ps.cmds, fmt.Sprintf(`
+$HashArguments = @{
+    CreateDNSDelegation           = $false
+    ForestMode                    = "Win2012R2"
+    DomainMode                    = "Win2012R2"
+    DomainName                    = "%s"
+    SafeModeAdministratorPassword = (ConvertTo-SecureString %s -AsPlainText -Force)
+    Force                         = $true
+}; Install-ADDSForest @HashArguments`, activeDirectoryDomain, passwd))
+	return ps
+}
+
+// UninstallADDSDomainController creates a command to remove a Domain Controller.
+func (ps *powerShellCommandBuilder) UninstallADDSDomainController(passwd string) *powerShellCommandBuilder {
+	ps.cmds = append(ps.cmds, fmt.Sprintf(`
+$HashArguments = @{
+	SkipPreChecks              = $true
+    LocalAdministratorPassword = (ConvertTo-SecureString %s -AsPlainText -Force)
+    DemoteOperationMasterRole  = $true
+	ForceRemoval               = $true
+	Force                      = $true
+}; Uninstall-ADDSDomainController
+`, passwd))
+	return ps
+}
+
+// AddActiveDirectoryUser creates a command for adding a user to an Active Directory domain.
+func (ps *powerShellCommandBuilder) AddActiveDirectoryUser(username, passwd string) *powerShellCommandBuilder {
+	ps.cmds = append(ps.cmds, fmt.Sprintf(`
+$HashArguments = @{
+	Name = '%s'
+	AccountPassword = (ConvertTo-SecureString %s -AsPlainText -Force)
+	Enabled = $true
+}; New-ADUser @HashArguments
+`, username, passwd))
+	return ps
+}
+
+// GetMachineType creates a command that returns the ProductType of the machine (2 for server, 3 for domain controller).
+func (ps *powerShellCommandBuilder) GetMachineType() *powerShellCommandBuilder {
+	ps.cmds = append(ps.cmds, "(Get-CimInstance -ClassName Win32_OperatingSystem).ProductType")
+	return ps
+}
+
+// StartService creates a command that starts a given service.
+func (ps *powerShellCommandBuilder) StartService(serviceName string) *powerShellCommandBuilder {
+	ps.cmds = append(ps.cmds, fmt.Sprintf(`
+Start-Service %s
+`, serviceName))
+	return ps
+}
+
+// WaitForServiceStatus creates a command that waits 1 minute for a service to reach a given state.
+func (ps *powerShellCommandBuilder) WaitForServiceStatus(serviceName, status string) *powerShellCommandBuilder {
+	// TODO: Make timeout configurable
+	ps.cmds = append(ps.cmds, fmt.Sprintf(`
+(Get-Service %s).WaitForStatus('%s', '00:01:00')
+`, serviceName, status))
+	return ps
+}
+
+// Execute compiles the list of PowerShell commands into one script and runs it on the given host
+func (ps *powerShellCommandBuilder) Execute(host *components.RemoteHost) (string, error) {
+	return host.Execute(ps.Compile())
+}
+
+// Compile joins all the saved command into one valid PowerShell script command.
+func (ps *powerShellCommandBuilder) Compile() string {
+	return strings.Join(ps.cmds, ";")
```

Comment:
by default PowerShell will continue on error. Since this is chaining commands together, if one of them fails the rest will still be executed. Maybe we can add support for setting [ErrorActionPreference](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_preference_variables?view=powershell-7.4#erroractionpreference) for the built command?